sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+  ThrowIfApiCallFails(pipe(m_fd.data()), "Cannot create output pipe");
+  ThrowIfApiCallFails(
+      fcntl(m_fd[0], F_SETFL, O_NONBLOCK), "Cannot set up output pipe to have non-blocking read");
+
+#endif
+}
+
+OutputPipe::~OutputPipe()
+{
+#if defined(AZ_PLATFORM_WINDOWS)
+  if (m_writeHandle != NULL)
+  {
+    static_cast<void>(CloseHandle(m_writeHandle));
+  }
+
+  if (m_readHandle != NULL)
+  {
+    static_cast<void>(CloseHandle(m_readHandle));
+  }
+#else
+  for (auto iter = m_fd.rbegin(); iter != m_fd.rend(); ++iter)
+  {
+    if (*iter != -1)
+    {
+      static_cast<void>(close(*iter));
+    }
+  }
+#endif
+}
+
+void EnsureShellExists(std::string const& pathToShell)
+{
+#ifdef _MSC_VER
+#pragma warning(push)
+// warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead.
+#pragma warning(disable : 4996)
+#endif
+
+  auto file = std::fopen(pathToShell.c_str(), "r");
+
+#ifdef _MSC_VER
+#pragma warning(pop)
+#endif
+
+  // LCOV_EXCL_START
+  if (!file)
+  {
+    throw std::runtime_error("Cannot locate command line shell.");
+  }
+  // LCOV_EXCL_STOP
+
+  std::fclose(file);
+}
+
+#if defined(AZ_PLATFORM_WINDOWS)
+void AppendToEnvironmentValuesIfNotEmpty(
+    std::vector<CHAR>& environmentValues,
+    std::string const& envVarName,
+    std::string const& value)
+{
+  if (!value.empty()) // LCOV_EXCL_LINE
+  {
+    auto const envVarStatement = envVarName + "=" + value;
+
+    environmentValues.insert(
+        environmentValues.end(), envVarStatement.begin(), envVarStatement.end());
+
+    environmentValues.push_back('\0'); // terminate the string
+  }
+}
+
+void AppendToEnvironmentValuesIfDefined(
+    std::vector<CHAR>& environmentValues,
+    std::string const& envVarName)
+{
+  AppendToEnvironmentValuesIfNotEmpty(
+      environmentValues, envVarName, Environment::GetVariable(envVarName.c_str()));
+}
+#else
+void AppendToArgvValues(
+    std::vector<char>& argvValues,
+    std::vector<std::remove_reference<decltype(argvValues)>::type::size_type>& argvValuePositions,
+    std::string const& value)
+{
+  argvValuePositions.push_back(argvValues.size());
+  argvValues.insert(argvValues.end(), value.begin(), value.end());
+  argvValues.push_back('\0');
+}
+#endif
+
+ShellProcess::ShellProcess(std::string const& command, OutputPipe& outputPipe)
+{
+#if defined(AZ_PLATFORM_WINDOWS)
+  // Start the process.
+  PROCESS_INFORMATION procInfo = {};
+  ZeroMemory(&procInfo, sizeof(decltype(procInfo)));
+
+  {
+    STARTUPINFO startupInfo = {};
+    ZeroMemory(&startupInfo, sizeof(decltype(startupInfo)));
+    startupInfo.cb = sizeof(decltype(startupInfo));
+    startupInfo.dwFlags |= STARTF_USESTDHANDLES; // cspell:disable-line
+    startupInfo.hStdInput = INVALID_HANDLE_VALUE;
+    startupInfo.hStdOutput = outputPipe.m_writeHandle;
+    startupInfo.hStdError = outputPipe.m_writeHandle;
+
+    // Path to cmd.exe
+    std::vector<CHAR> commandLineStr;
+    {
+      auto cmd = Environment::GetVariable("COMSPEC");
```

Comment:
Why are you verifying that COMSPEC exists? For that matter, on Windows, you should just assume that you can launch a command named "cmd /c" always.

Reply:
I need the full path, not just the name. `COMSPEC` contains full path.

I need the full path to executable to start the process.

Reply:
My point is that you *don't* need the full path, if you're running on Windows or Linux, since the OS searches the PATH environment to find the executable. And sh and cmd are always in the path.

Reply:
I think I even found the way to do it using POSIX API - via `posix_spawnp()`, but for Win32, I can't make it to work as documentation describes.